wright/src/grammar/parsers/expression/binary_expression.rs

```diff
@@ -13,9 +65,86 @@ impl<'s> BinaryExpression<'s> {
         }
     }
 
+    fn dot(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Expression::primary,
+            &[BinaryOp::Dot],
+        )(input)
+    }
+
+    fn dotdot(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Self::dot,
+            &[BinaryOp::DotDot],
+        )(input)
+    }
+
+    fn mul(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Self::dotdot,
+            &[BinaryOp::Mul, BinaryOp::Div, BinaryOp::Mod],
+        )(input)
+    }
+
+    fn add(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Self::mul,
+            &[BinaryOp::Add, BinaryOp::Sub, BinaryOp::Mod],
+        )(input)
+    }
+
+    fn and(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Self::add,
+            &[BinaryOp::And],
+        )(input)
+    }
+
+    fn xor(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Self::and,
+            &[BinaryOp::Xor],
+        )(input)
+    }
+
+    fn or(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Self::xor,
+            &[BinaryOp::Or],
+        )(input)
+    }
+
+    fn andand(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Self::or,
+            &[BinaryOp::AndAnd],
+        )(input)
+    }
+
+    fn oror(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Self::andand,
+            &[BinaryOp::OrOr],
+        )(input)
+    }
+
+    fn cmp(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Self::oror,
+            &[BinaryOp::Lt, BinaryOp::Gt, BinaryOp::Le, BinaryOp::Ge],
+        )(input)
+    }
+
+    fn eq(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        binary_parser(
+            Self::cmp,
+            &[BinaryOp::EqEq, BinaryOp::NotEq],
+        )(input)
+    }
+
     /// Parse a binary expression in source code.
-    pub fn parse(input: Fragment<'s>) -> IResult<Fragment<'s>, Self> {
-        todo!("binary expression parser")
+    pub fn parse(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        Self::eq(input)
```

Comment:
Binary expression parsing should return a `BinaryExpression`, not an `Expression`.

Reply:
Ready for second review